demo/KevinDemo/Networking/Base/BaseApiClient.swift

```diff
@@ -38,33 +37,39 @@ open class PSBaseApiClient {
         session.cancelAllRequests()
     }
     
-    public func doRequest<RC: URLRequestConvertible, E: Mappable>(requestRouter: RC) -> Promise<[E]> {
+    public func doRequest<RC: URLRequestConvertible, E: Codable>(requestRouter: RC) -> Promise<[E]> {
         let request = createRequest(requestRouter)
         executeRequest(request)
         
         return request
             .pendingPromise
             .promise
             .map(on: workQueue) { body in
-                guard let objects = Mapper<E>().mapArray(JSONObject: body) else {
+                do {
```

Comment:
May I suggest a new class called ResponseDecoder that would look something like this 

```
    class ResponseDecoder {
        
        private let decoder = JSONDecoder()
        
        func decodeRequest<E: Codable>(with body: Any) throws -> E {
            let data = try makeData(with: body)
            do {
                return try decoder.decode(E.self, from: data)
            } catch {
                throw makeError(with: data)
            }
        }
        
        private func makeData(with body: Any) throws -> Data {
            do {
                return try JSONSerialization.data(withJSONObject: body, options: [])
            } catch {
                throw ApiError.unknown()
            }
        }
        
        private func makeError(with data: Data) -> ApiError {
            let error = try? decoder.decode(ApiError.self, from: data)
            return error ?? .unknown()
        }
    }
```

This way we can have here something like this

```
            .compactMap(on: workQueue) { [weak self] body in
                try self?.responseDecoder.decodeRequest(with: body)
            }
```